packages/near-membrane-base/src/environment.ts

```diff
@@ -1,192 +1,181 @@
-            construct(
-                target: BlueConstructor,
-                argumentsList: ArrayLike<BlueValue>,
-                newTarget?: any
-                // eslint-disable-next-line class-methods-use-this
-            ): BlueValue {
-                return ReflectConstruct(target, argumentsList, newTarget);
-            },
+        const { blueConnector: localInit, redConnector: foreignInit, distortionCallback } = options;
+
+        let blueHooks: Parameters<HooksCallback>;
+        let redHooks: Parameters<HooksCallback>;
+
+        const blueExportsCallback: HooksCallback = (...hooks) => {
+            blueHooks = hooks;
+        };
+        const redExportsCallback: HooksCallback = (...hooks) => {
+            redHooks = hooks;
         };
-        this.getRedValue = redEnvFactory(this, blueHooks);
-        this.getBlueValue = blueProxyFactory(this);
-    }
 
-    // istanbul ignore next
-    // eslint-disable-next-line class-methods-use-this
-    getBlueValue(
-        // eslint-disable-next-line @typescript-eslint/no-unused-vars
-        red: RedValue
-    ): BlueValue {
-        // placeholder since this will be assigned in construction
-    }
+        const localConnect = localInit(
+            undefinedSymbol,
+            'blue',
+            false,
+            blueExportsCallback,
+            distortionCallback
+        );
+        const foreignConnect = foreignInit(undefinedSymbol, 'red', true, redExportsCallback);
+        localConnect(...redHooks!);
+        foreignConnect(...blueHooks!);
 
-    // istanbul ignore next
-    // eslint-disable-next-line class-methods-use-this
-    getRedValue(
-        // eslint-disable-next-line @typescript-eslint/no-unused-vars
-        blue: BlueValue
-    ): RedValue {
-        // placeholder since this will be assigned in construction
+        const [blueGetSelectedTarget, blueGetTransferableValue] = blueHooks!;
+        this.blueGetSelectedTarget = blueGetSelectedTarget;
+        this.blueGetTransferableValue = blueGetTransferableValue;
+        // prettier-ignore
+        const [,,
+            redCallableEvaluate,
+            redCallableInstallLazyDescriptors,,,,
+            redCallableDefineProperty,,,,,,,,
+            redCallableSetPrototypeOf
+        ] = redHooks!;
+        this.redCallableEvaluate = redCallableEvaluate;
+        this.redCallableInstallLazyDescriptors = redCallableInstallLazyDescriptors;
+        this.redCallableSetPrototypeOf = redCallableSetPrototypeOf;
+        this.redCallableDefineProperty = redCallableDefineProperty;
     }
 
-    setRefMapEntries(red: RedObject, blue: BlueObject) {
-        // double index for perf
-        WeakMapSet(this.redMap, red, blue);
-        WeakMapSet(this.blueMap, blue, red);
+    evaluate(sourceText: string): void {
+        try {
+            this.redCallableEvaluate(sourceText);
+        } catch (e) {
+            const pushedError = this.blueGetSelectedTarget();
+            if (pushedError !== undefined) {
```

Comment:
if `blueSelectedTarget` is always an object or nothing we can do a truthy check here as just `if (pushedError) {`